src/core/settings.c

```diff
@@ -661,24 +662,44 @@ void sig_term(int n)
 	raise(SIGTERM);
 }
 
-/* Yes, this is my own stupid checksum generator, some "real" algorithm
-   would be nice but would just take more space without much real benefit */
-static unsigned int file_checksum(const char *fname)
+static int file_checksum(const char *fname, const guint8 *old_checksum, gsize old_checksum_length,
+                         guint8 **checksum_out, gsize *checksum_length_out)
 {
-        char buf[512];
-        int f, ret, n;
-	unsigned int checksum = 0;
+	GChecksum *c;
+	unsigned char buf[512];
+	guint8 *checksum;
+	gsize checksum_length;
+	int f, ret;
+	static GChecksumType checksum_type = G_CHECKSUM_SHA1;
+
+	if ((c = g_checksum_new(checksum_type)) == NULL) {
+		return FALSE;
+	}
 
 	f = open(fname, O_RDONLY);
-	if (f == -1) return 0;
+	if (f == -1)
+		return FALSE;
 
-        n = 0;
 	while ((ret = read(f, buf, sizeof(buf))) > 0) {
-		while (ret-- > 0)
-			checksum += buf[ret] << ((n++ & 3)*8);
+		g_checksum_update(c, buf, ret);
 	}
 	close(f);
-	return checksum;
+
+	checksum = g_new0(guint8, g_checksum_type_get_length(checksum_type));
+	g_checksum_get_digest(c, checksum, &checksum_length);
```

Comment:
This crashes with valgrind (and gets good enough garbage in the stack to work decently the rest of the time)

```
==28250== Conditional jump or move depends on uninitialised value(s)
==28250==    at 0x56A2855: g_checksum_get_digest (in /usr/lib/libglib-2.0.so.0.5400.0)
==28250==    by 0x1B1AAE: file_checksum (settings.c:689)
==28250==    by 0x1B1BC1: irssi_config_save_state (settings.c:717)
==28250==    by 0x1B1C57: parse_configfile (settings.c:774)
==28250==    by 0x1B3310: init_configfile (settings.c:794)
==28250==    by 0x1B3310: settings_init (settings.c:901)
==28250==    by 0x19FE44: core_init (core.c:237)
==28250==    by 0x12BE0F: textui_init (irssi.c:147)
==28250==    by 0x12BE0F: main (irssi.c:310)

(irssi:28250): GLib-CRITICAL **: g_checksum_get_digest: assertion '*digest_len >= len' failed
zsh: trace trap (core dumped)  valgrind --log-file=leg ./src/fe-text/irssi
```

checksum_length is inout, so initialize it to `g_checksum_type_get_length(checksum_type)`

Reply:
great